src/hotspot/share/runtime/thread.cpp

```diff
@@ -2456,24 +2456,32 @@ size_t      JavaThread::_stack_size_at_create = 0;
 bool        Threads::_vm_complete = false;
 #endif
 
-static inline void *prefetch_and_load_ptr(void **addr, intx prefetch_interval) {
+static inline JavaThread *prefetch_and_load_ptr(JavaThread*const * addr, intx prefetch_interval) {
   Prefetch::read((void*)addr, prefetch_interval);
   return *addr;
 }
 
-// Possibly the ugliest for loop the world has seen. C++ does not allow
-// multiple types in the declaration section of the for loop. In this case
-// we are only dealing with pointers and hence can cast them. It looks ugly
-// but macros are ugly and therefore it's fine to make things absurdly ugly.
-#define DO_JAVA_THREADS(LIST, X)                                                                                          \
-    for (JavaThread *MACRO_scan_interval = (JavaThread*)(uintptr_t)PrefetchScanIntervalInBytes,                           \
-             *MACRO_list = (JavaThread*)(LIST),                                                                           \
-             **MACRO_end = ((JavaThread**)((ThreadsList*)MACRO_list)->threads()) + ((ThreadsList*)MACRO_list)->length(),  \
-             **MACRO_current_p = (JavaThread**)((ThreadsList*)MACRO_list)->threads(),                                     \
-             *X = (JavaThread*)prefetch_and_load_ptr((void**)MACRO_current_p, (intx)MACRO_scan_interval);                 \
-         MACRO_current_p != MACRO_end;                                                                                    \
-         MACRO_current_p++,                                                                                               \
-             X = (JavaThread*)prefetch_and_load_ptr((void**)MACRO_current_p, (intx)MACRO_scan_interval))
+struct JavaThreadPrefetchedIterator {
+    ThreadsList* _list;
+    JavaThread*const * _end;
+    JavaThread*const * _current;
+
+    JavaThreadPrefetchedIterator(ThreadsList* list) :
+      _list(list), _end(list->threads() + list->length()), _current(list->threads()) {}
+
+    JavaThread* current() {
+      return _current != _end
+        ? prefetch_and_load_ptr(_current, PrefetchScanIntervalInBytes)
+        : NULL;
+    }
+
+    void next() {
+      _current++;
+    }
+  };
+
+#define DO_JAVA_THREADS(LIST, X) \
+  for (JavaThreadPrefetchedIterator iter(LIST); JavaThread* X = iter.current(); iter.next())
```

Comment:
The condition is an implicit conversion to bool, which is disallowed
by the style guide.  In fact, it's a declaration, which is
specifically called out by the style guide.  When I revised the style
guide I made explicit mention of that usage because I'd tried to use
that sort of thing several years ago and got several complaints.  So
it's not that I personally think that usage is bad; rather, I was
hoping someone would take on the little project of convincing the team
it was good.  I don't have any objection to this part of the change.

Reply:
I can't find a way to fix that without adding back some of the crazy things
in the original macro.

Reply:
I wasn't expecting any change based on that comment, just pointing out that the code
*is* contrary to the style guide, in a place where I think the guide should be changed.

Reply:
I'm surprised that a declaration is even legal in that position - but then I should never be surprised at the crazy things C++ allows. :( I do think  it looks truly awful to have a declaration there. Surely we do not have to reduce this to being a single-line macro that emulates a for-loop? That said I can't make any sense out of the original macro so sure this is a less-ugly macro. But why does this have to be so obscure - aren't we just walking the current threads-list ??

Reply:
I failed to notice previously that the new macro introduces the "iter" variable into the body scope.  The old macro used naming conventions to avoid adding any "simple" names that could inadvertently affect the meaning of the code.

Thinking about David's reply, I'm also wondering if it might be possible to do better by making use of range-based for.  It would involve defining a compliant iterator rather than the JavaThreadPrefetchedIterator, but that's not very hard.  (And I can help with that.)  Doing so would really simplify the macro, or perhaps even render it so trivial as to not be worth using any more.

So I'm rescinding my approval of this change.